cmd/manager/main.go

```diff
@@ -611,8 +614,20 @@ func asyncTasks(
 		}()
 	}
 
-	// Garbage collect any orphaned user Secrets leftover from deleted resources while the operator was not running.
+	// Garbage collect orphan secrets leftover from deleted resources while the operator was not running
+	// - association user secrets
 	garbageCollectUsers(cfg, managedNamespaces)
+	// - soft-owned secrets
+	if err := reconciler.GarbageCollectAllOrphanSecrets(k8s.WrapClient(mgr.GetClient()), map[string]runtime.Object{
+		esv1.Kind:        &esv1.Elasticsearch{},
+		apmv1.Kind:       &apmv1.ApmServer{},
+		kbv1.Kind:        &kbv1.Kibana{},
+		entv1beta1.Kind:  &entv1beta1.EnterpriseSearch{},
+		beatv1beta1.Kind: &beatv1beta1.Beat{},
```

Comment:
I'm not super happy with that :( 
If anyone has another suggestion.

Reasoning behind this:
- when reconciling soft-owned secrets, we need to set a label representing the Kind ("Elasticsearch"), to not mismatch Elasticsearch with Kibana for example. The Kind can just be retrieved from the given owner `runtime.Object`, which is great.
- when running the garbage collection at startup, we need to `client.Get()` owners, which requires a typed object in parameter (e.g. `&esv1.Elasticsearch{}`)
- this map matches the soft owner Kind label with an actual object type for the garbage collection function
- retrieving the Kind from an apiserver object works fine, but `&esv1.Elasticsearch{}.Kind` is empty because the object did not go through the apiserver. I realized the operator bootstrap code actually performs reflection in `SchemeBuilder.Register()` to come up with that value...

Reply:
What are you not happy with? The fact that you have to spec both `Kind` and the corresponding struct somewhat over-specifying even though one could be derived from the other? 

The only other idea how to potentially avoid this was not to use the controller-runtime client but using the [dynamic client from client-go directly](https://github.com/kubernetes/client-go/blob/v0.18.1/dynamic/interface.go#L39), in which case `Kind` should suffice and as we only care about exists/does not exist we would not need to worry about REST mapping etc. But it feels like a lot of trouble for little gain.

Reply:
Yeah 👍 
Using our usual k8s.Client makes it easier to test, so I think I'd rather keep things as they are.